### PR TITLE
Change to pull image by default

### PR DESCRIPTION
**Pull request by JamesJeffryes**
I also did a bit of code cleanup: fixed some typos in what must be more of less dead code, removed un used imports and extra whitespace

### CODE REVIEW

File: lib/biokbase/catalog/registrar.py [hunk 1]
```diff
@@ -437,15 +434,15 @@ def validate_method_specs(self, basedir):
                         display_yaml = display_yaml_file.read()
 
                     # gather any extra html files
-                    extraFiles = {}
+                    extra_files = {}
```
Comment (MrCreosote) [inline lib/biokbase/catalog/registrar.py]: If this really is dead code it should be deleted. If it's not dead code and we're changing it we should test it.

Comment (MrCreosote) [reply lib/biokbase/catalog/registrar.py]: this section is an argument for Java :P

File: lib/biokbase/catalog/registrar.py [hunk 2]
```diff
@@ -546,7 +536,7 @@ def build_docker_image(self, docker_client, image_name, basedir):
 
         # examine stream to determine success/failure of build
         imageId = None
-        for lines in docker_client.build(path=basedir, rm=True, tag=image_name, pull=False):
+        for lines in docker_client.build(path=basedir, rm=True, tag=image_name, pull=True):
```
Comment (MrCreosote) [inline lib/biokbase/catalog/registrar.py]: Will this cause problems for the old sdkbase image that shouldn't get pulled?

Comment (MrCreosote) [reply lib/biokbase/catalog/registrar.py]: Also is there any way to test this?

**Comment by MrCreosote** [comment]
Also tests are all failing

**Comment by JamesJeffryes** [comment]
I think there may be an issue with the travis due to the auth service upgrade. The master build is failing with these errors `Exception: {"error":{"httpcode":400,"httpstatus":"Bad Request","appcode":30000,"apperror":"Missing input parameter","message":"30000 Missing input parameter: token","callid":"1742835537693122","time":1507679329844}}`

I'm not sure why the password is not getting correctly set by travis in my PR but I decided to press on with looking into switching to auth2

When add the CI Auth2 URL and a token from CI I get BUILD_ERROR: JSONRPCError: -32400. Token validation failed: Auth token is in the incorrect format, near <token>
us.kbase.auth.TokenFormatException: Auth token is in the incorrect format, near <token>
	at us.kbase.auth.AuthToken.<init>(AuthToken.java:90)
	at us.kbase.auth.AuthToken.<init>(AuthToken.java:46)
	at us.kbase.common.service.JsonServerServlet.validateToken(JsonServerServlet.java:491)
	at us.kbase.common.service.JsonServerServlet.doPost(JsonServerServlet.java:338)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:530)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:426)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:118)
	at org.eclipse.jetty.server.session.SessionHandler.handle(SessionHandler.java:179)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:930)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:361)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:864)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:116)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:112)
	at org.eclipse.jetty.server.Server.handle(Server.java:333)
	at org.eclipse.jetty.server.HttpConnection.handleRequest(HttpConnection.java:552)
	at org.eclipse.jetty.server.HttpConnection$RequestHandler.content(HttpConnection.java:993)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:739)
	at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:209)
	at org.eclipse.jetty.server.HttpConnection.handle(HttpConnection.java:399)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.run(SelectChannelEndPoint.java:437)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:435)
	at java.lang.Thread.run(Thread.java:748)

I'm sure the token is being copied correctly because the <token> in the message above matches my dev token

File: lib/biokbase/catalog/registrar.py (hunk 2 above)
Comment (JamesJeffryes) [reply lib/biokbase/catalog/registrar.py]: The only time that might cause trouble is if the sdkbase image on docker is newer than the sdkbase image in the environment running the catalog. (I'd have to ask Keith to get that listing for each environment). The new images that we have been building are different namespace so should be OK.

I should be able to test this by looking that the docker image is not cached in the build log. To do that I need tests on the master branch to work however.

File: lib/biokbase/catalog/registrar.py (hunk 1 above)
Comment (JamesJeffryes) [reply lib/biokbase/catalog/registrar.py]: So, this looks to me to support adding additional html files to a method for extra documentation. We probably never used the functionality so never noticed it was broken. I can test by adding an extra html file to https://github.com/kbaseIncubator/catalog_test_module/tree/simple_good_repo/ui/narrative/methods/test_method_1
But I'm not sure it's even worth the effort.

Comment (MrCreosote) [reply lib/biokbase/catalog/registrar.py]: Hmm, I could've sworn that some modules had extra html files. Maybe it worked at one point?

If you want to remove it, we should check whether a) any modules have extra html files and b) whether we still need to support the feature.

File: lib/biokbase/catalog/registrar.py (hunk 2 above)
Comment (MrCreosote) [reply lib/biokbase/catalog/registrar.py]: > The only time that might cause trouble is if the sdkbase image on docker is newer than the sdkbase image in the environment running the catalog. (I'd have to ask Keith to get that listing for each environment)

That's what I'm concerned about - if we're not careful we could bork the sdkbase image.

**Comment by MrCreosote** [comment]
> I'm not sure why the password is not getting correctly set by travis in my PR

A password is not going to work - the tests won't go to google or wherever to log in. Also, any encrypted travis info will not be available in PRs, so given the tests require a token, there's no way to get them to pass in a PR currently. 

The fix would be to use the auth service via mini KBase or just locally to handle the tokens, and then tokens can be created on the fly.

**Comment by MrCreosote** [comment]
That stack trace is coming from an old auth client that expects the Globus token format - what's throwing that error?

File: lib/biokbase/catalog/registrar.py (hunk 1 above)
Comment (JamesJeffryes) [reply lib/biokbase/catalog/registrar.py]: So I added a trivial test that registers a module with extra html files just to make sure that nothing breaks. I can't really figure out what NMS does with the extra files but I'll argue that those tests belong in NMS not here.

**Comment by JamesJeffryes** [comment]
In order for these tests to pass on Travis, we need to register a long term token as a secure parameter. or figure out a way to make this work in mini kbase.

They do however pass locally for me

File: lib/biokbase/catalog/registrar.py (hunk 2 above)
Comment (JamesJeffryes) [reply lib/biokbase/catalog/registrar.py]: The sdkbase on docker hub is over a year old.  https://hub.docker.com/r/kbase/sdkbase/tags/

Keith sez:
>on prod, the kbase/kbase:sdkbase.latest is 10 months old, so it's probably v1
>
>(there is a kbase/sdkbase2:latest image there too)
>on ci, there's kbase/kbase:sdkbase2.latest and kbase/kbase:sdkbase.latest.  the latter is 6 >weeks old

The versions on the server are fresher than docker hub so I think we should be OK

Comment (MrCreosote) [reply lib/biokbase/catalog/registrar.py]: @sychan you're sdkbase docker image expert at this point, what say you?

**Comment by MrCreosote** [comment]
> we need to register a long term token as a secure parameter
That still won't work for PRs from forks

File: lib/biokbase/catalog/Impl.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def __init__(self, config):
         #BEGIN_CONSTRUCTOR
         print('Starting the Catalog service.  Service configuration:')
         for c in config:
-            if c == 'nms-admin-psswd':
+            if c == 'nms-admin-psswd' or c == 'nms-admin-token':
```
Comment (MrCreosote) [inline lib/biokbase/catalog/Impl.py]: If you want you could just strip out the user/pwd config items. They're obsolete.

File: lib/biokbase/catalog/registrar.py [hunk 4]
```diff
@@ -461,18 +460,13 @@ def validate_method_specs(self, basedir):
                                 for e in result['errors']:
                                     self.log('        - error: '+e, is_error=True)
                         else:
-                            self.log('        - error is undefined!'+e,  is_error=True)
```
Comment (MrCreosote) [inline lib/biokbase/catalog/registrar.py]: wow

File: lib/biokbase/catalog/baseclient.py [hunk 5]
```diff
@@ -121,7 +121,7 @@ def __init__(
             self, url=None, timeout=30 * 60, user_id=None,
             password=None, token=None, ignore_authrc=False,
             trust_all_ssl_certificates=False,
-            auth_svc='https://kbase.us/services/authorization/Sessions/Login',
+            auth_svc='https://ci.kbase.us/services/auth/api/legacy/KBase/Sessions/Login',
```
Comment (MrCreosote) [inline lib/biokbase/catalog/baseclient.py]: This is generated code - did you regenerate or edit?

File: lib/biokbase/narrative_method_store/client.py [hunk 6]
```diff
@@ -25,8 +25,7 @@
 
 
 def _get_token(user_id, password,
-               auth_svc='https://nexus.api.globusonline.org/goauth/token?' +
-                        'grant_type=client_credentials'):
+               auth_svc="https://ci.kbase.us/services/auth/api/legacy/KBase/Sessions/Login"):
```
Comment (MrCreosote) [inline lib/biokbase/narrative_method_store/client.py]: This is generated code - did you regenerate or edit?

File: test/core_registration_test.py [hunk 7]
```diff
@@ -655,7 +655,29 @@ def test_active_inactive_remove_module(self):
         self.assertEqual(str(e.exception),
             'Cannot delete module that has been released.  Make it inactive instead.');
 
+    def test_extra_files(self):
+        # (1) register the test repo
+        giturl = self.cUtil.get_test_repo_1()
```
Comment (MrCreosote) [inline test/core_registration_test.py]: Where does cUtil come from?

File: lib/biokbase/catalog/registrar.py [hunk 8]
```diff
@@ -437,16 +434,18 @@ def validate_method_specs(self, basedir):
...
 
                     # gather any extra html files
-                    extraFiles = {}
+                    extra_files = {}
                     for extra_file_name in os.listdir(os.path.join(method_path)):
-                        if not os.path.isfile(os.path.join(method_path,extra_file_name)): break
-                        if not extra_file_name.endswith('.html'): break
-                        with codecs.open(oos.path.join(method_path,extra_file_name), 'r', "utf-8", errors='ignore') as extra_file: 
-                            extrafiles[extra_file_name] = extra_file.read()
+                        if not os.path.isfile(os.path.join(method_path,extra_file_name)):
```
Comment (MrCreosote) [inline lib/biokbase/catalog/registrar.py]: The tests just ensure that the extra file doesn't cause the registration to fail. That doesn't really test what's going on here or the 3 paths through the code (or the loop)

File: test/core_registration_test.py (hunk 7 above)
Comment (JamesJeffryes) [reply test/core_registration_test.py]: It's defined in the setup class L669

File: lib/biokbase/catalog/registrar.py (hunk 8 above)
Comment (JamesJeffryes) [reply lib/biokbase/catalog/registrar.py]: So the 'if not extra_file_name.endswith('.html'):' gets triggered by the spec.json and display.yaml files but I'll add a directory to the test repo to make sure the last path gets triggered

As I said above, I don't really understand what NMS is doing with the passed files other than storing them and I think testing that functionality should be in NMS tests not here in catalog

File: lib/biokbase/catalog/Impl.py (hunk 3 above)
Comment (JamesJeffryes) [reply lib/biokbase/catalog/Impl.py]: Done

File: lib/biokbase/catalog/baseclient.py (hunk 5 above)
Comment (JamesJeffryes) [reply lib/biokbase/catalog/baseclient.py]: Edited. Reverting this change.

File: lib/biokbase/narrative_method_store/client.py (hunk 6 above)
Comment (JamesJeffryes) [reply lib/biokbase/narrative_method_store/client.py]: Edited. Copied new generated code from NMS

File: lib/biokbase/catalog/registrar.py (hunk 8 above)
Comment (MrCreosote) [reply lib/biokbase/catalog/registrar.py]: I could remove this code entirely and the test would still pass. That means the test is not effective.

File: test/test.cfg.example [hunk 9]
```diff
@@ -41,8 +41,6 @@ docker-push-allow-insecure = 0
 # Narrative Method Store configuration. Provide either a token or a uid/pwd.
```
Comment (MrCreosote) [inline test/test.cfg.example]: Remove the the 'or' part

File: lib/biokbase/catalog/baseclient.py [hunk 10]
```diff
@@ -121,7 +121,7 @@ def __init__(
             self, url=None, timeout=30 * 60, user_id=None,
             password=None, token=None, ignore_authrc=False,
             trust_all_ssl_certificates=False,
-            auth_svc='https://ci.kbase.us/services/auth/api/legacy/KBase/Sessions/Login',
+            auth_svc='https://kbase.us/services/authorization/Sessions/Login',
```
Comment (MrCreosote) [inline lib/biokbase/catalog/baseclient.py]: We should update the SDK template for the new url...

File: test/core_registration_test.py (hunk 7 above)
Comment (MrCreosote) [reply test/core_registration_test.py]: Sorry, I'm having trouble figuring this out. What setup class?

**Comment by MrCreosote** [comment]
Also will need release notes

File: test/core_registration_test.py (hunk 7 above)
Comment (JamesJeffryes) [reply test/core_registration_test.py]: https://github.com/JamesJeffryes/catalog/blob/f435e20abf6450c08a43b629d7bee2258c52861b/test/core_registration_test.py#L691

File: lib/biokbase/catalog/baseclient.py (hunk 10 above)
Comment (JamesJeffryes) [reply lib/biokbase/catalog/baseclient.py]: Done in py3 fork

File: lib/biokbase/catalog/registrar.py (hunk 8 above)
Comment (JamesJeffryes) [reply lib/biokbase/catalog/registrar.py]: I don't know how to test this any better, maybe we just strip it out. NMS just swallows up the extra_files without a peep. Nothing I can see in the NMS spec even returns "extra_files"

File: test/test.cfg.example (hunk 9 above)
Comment (JamesJeffryes) [reply test/test.cfg.example]: Sure

File: lib/biokbase/catalog/registrar.py (hunk 8 above)
Comment (MrCreosote) [reply lib/biokbase/catalog/registrar.py]: It looks to me like the validate_method() call that takes the extra_files parameter doesn't actually make any changes NMS side - it just validates that everything is OK. The extra files end up being used in this 700+ line method: https://github.com/kbase/narrative_method_store/blob/master/src/us/kbase/narrativemethodstore/db/NarrativeMethodData.java#L89-L801

Hence, the only way to really test this is to make it fail based on the content of the extra files, as far as I can tell. 

Given you would have to figure out what that 700+ line method does and how to make it fail with some magic file, I think it's time to give up.

Comment (JamesJeffryes) [reply lib/biokbase/catalog/registrar.py]: So strip out the python functionality?

Comment (MrCreosote) [reply lib/biokbase/catalog/registrar.py]: Leave it I suppose.

File: README.md [hunk 11]
```diff
@@ -11,6 +11,11 @@ Code coverage: (develop branch)
 [![Coverage Status](https://coveralls.io/repos/github/kbase/catalog/badge.svg?branch=develop)](https://coveralls.io/github/kbase/catalog?branch=develop)
 
 
+#### v2.1.2 - 3/16/18
+  - Pull a new base image if possible each time a module is registered
+  - Fix extra_files logic
+  - Update NMS to use Auth2
```
Comment (MrCreosote) [inline README.md]: Since this only affects testing, I don't think it should be included in release notes - confusing otherwise

File: README.md [hunk 12]
```diff
@@ -11,6 +11,11 @@ Code coverage: (develop branch)
 [![Coverage Status](https://coveralls.io/repos/github/kbase/catalog/badge.svg?branch=develop)](https://coveralls.io/github/kbase/catalog?branch=develop)
 
 
+#### v2.1.2 - 3/16/18
+  - Pull a new base image if possible each time a module is registered
+  - Fix extra_files logic
```
Comment (MrCreosote) [inline README.md]: Needs to be more specific about what this means and where

File: README.md [hunk 13]
```diff
@@ -13,8 +13,8 @@ Code coverage: (develop branch)
 
 #### v2.1.2 - 3/16/18
   - Pull a new base image if possible each time a module is registered
-  - Fix extra_files logic
-  - Update NMS to use Auth2
+  - Fix the logic that allows additional html files to be passed from a method's
```
Comment (MrCreosote) [inline README.md]: Sorry for being annoying about this, but we should say it's specifically for the validation step. There may be another step later that actually changes the backend DB after the image is built.

Comment (JamesJeffryes) [reply README.md]: Done

File: lib/biokbase/catalog/registrar.py (hunk 2 above)
Comment (MrCreosote) [reply lib/biokbase/catalog/registrar.py]: > I should be able to test this by looking that the docker image is not cached in the build log. 

So is it possible to add these tests now?

**Comment by MrCreosote** [comment]
Other than the question of how to test the base image pull and @sychan 's blessing, I think  this is ready.

**Comment by sychan** [comment]
I also noticed that the version number doesn't seem to be updated, at least not in the impl file.

What was it that you thought needed my attention?

**Comment by JamesJeffryes** [comment]
@sychan When you have a chance, can you take a look at this again and consider merging? I addressed the version numbers from you comment and responded to Gavin's comments

**Comment by sychan** [comment]
You can't register travis secure tokens in one repository and expect it to work across repositories, but you can definitely register separate tokens in each repository via travis.

**Pull request merged**